### PR TITLE
Sendgrid - new actions

### LINKED ISSUE #1248: [ACTION] Support SendGrid Actions
**Is there a specific app this action is for?**
Sendgrid

Add Emails to Global Suppression | Adds recipient email address to the global suppression group.
Add or Update a Contact |	Creates new contact or updates existing based on the email address.
Add or Update a Recipient (Legacy) | Adds a new recipient or updates an existing one based on the email address. 
Add Recipients to a List (Legacy) |	Adds existing recipients to a list. 
Create a List | Creates a new list.
Delete a Global Suppression | Removes an email address from the global suppressions group.
Delete a List |	Deletes a list.
Delete all Bounces | Deletes all bounces.
Delete Blocks |	Deletes either all or specific emails on your block list.
Delete Bounces | Deletes specified bounces.
Delete Contacts | Deletes one or more contacts.
Delete Recipient from a List (Legacy)	| Deletes recipient from a specific list.
Delete Recipients (Legacy) | Deletes one or more recipients.
Get a Block	| Retrieves a specific email address from your block list.
Get a Global Suppression	| Retrieves a specific global suppression email.
Make an API Call | Performs an arbitrary authorized API call.
Remove Contacts from a List	| Removes contacts from the given list. The contacts will not be deleted. 
Send an Email	| Sends an email.
Validate an Email	| Validates an email address.
Get all Bounces	| Retrieves all of your bounces.
Get all Contacts | Retrieves all your contacts.
Get all Lists	| Retrieves all your contact lists.
Get all Recipients | Retrieves all of your recipients or recipients under a specified list.
List Blocks	| Lists all email addresses that are currently on your block list.
List Global Suppressions | Lists all email addresses that are globally suppressed.
Search Contacts | Searches for contacts. 

**Please provide a link to the relevant API docs for the specific service / operation.**
https://sendgrid.com/docs/api-reference/
For Legacy endpoints: https://www.twilio.com/docs/sendgrid/api/v2/using_the_web_api

### PR DESCRIPTION
**Pull request by michelle0927**
This PR replaces https://github.com/PipedreamHQ/pipedream/pull/1288

Resolves #1248

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/sendgrid/actions/get-a-block/get-a-block.mjs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+export default {
+  ...common,
+  key: "sendgrid-get-a-block",
+  name: "Get a Block",
+  description: "Gets a specific block. [See the docs here](https://docs.sendgrid.com/api-reference/blocks-api/retrieve-a-specific-block)",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    ...common.props,
+    email: {
```
Comment (vellames-turing) [inline components/sendgrid/actions/get-a-block/get-a-block.mjs:13]: I notice some repeatable uses of this variable through the actions. Maybe we can add it into `app`?

File: components/sendgrid/actions/list-blocks/list-blocks.mjs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+export default {
+  ...common,
+  key: "sendgrid-list-blocks",
+  name: "List Blocks",
+  description: "Allows you to list all email addresses that are currently on your blocks list. [See the docs here](https://docs.sendgrid.com/api-reference/blocks-api/retrieve-all-blocks)",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    ...common.props,
+    startTime: {
```
Comment (vellames-turing) [inline components/sendgrid/actions/list-blocks/list-blocks.mjs:13]: I notice some repeatable uses of this variable through the actions. Maybe we can add it into `app`?

File: components/sendgrid/actions/list-blocks/list-blocks.mjs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      type: "integer",
+      label: "Start Time",
+      description: "The start of the time range when a blocked email was created (inclusive). This is a unix timestamp.",
+      optional: true,
+    },
+    endTime: {
```
Comment (vellames-turing) [inline components/sendgrid/actions/list-blocks/list-blocks.mjs:19]: I notice some repeatable uses of this variable through the actions. Maybe we can add it into `app`?

**Review by vellames-turing — DISMISSED** [review]
Looks really great to me.
I saw some variables being repeated across the actions. But since they are not async options and the description changes, I am approving even with these repeatable props.
What do you think about moving these repeatable to `app`?

**Review by vellames-turing — DISMISSED** [review]

**Comment by dannyroosevelt** [comment]
Ran into this issue when trying to list `From Email` in the `send_email_single_recipient` action: 
image

File: components/sendgrid/sendgrid.app.mjs [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+    },
+    asm: {
+      type: "object",
+      label: "ASM",
+      description: "Advanced Suppression Manager. An object allowing you to specify how to handle unsubscribes",
+      optional: true,
+    },
+    ipPoolName: {
+      type: "string",
+      label: "Ip Pool Name",
```
Comment (dannyroosevelt) [inline components/sendgrid/sendgrid.app.mjs:155]: should be "IP Pool Name"

**Comment by dannyroosevelt** [comment]
The structure of the `content` prop for the `send_email_single_recipient` action is also pretty confusing -- do we need to have users construct an object there? How can we simplify it?

**Comment by dannyroosevelt** [comment]
> @dannyroosevelt I updated the `content` prop to just contain a string with the email content in text and/or html, which I think is clearer.
> 
> I'm able to reproduce that error in `From Email` if I change the API Key Permissions in my Sendgrid account. Can you check and see if your API Key has permissions set to Full Access? Screen Shot 2022-05-19 at 11 52 14 AM

Cool, thanks for updating the `content` prop!

Re: API permissions, that makes sense -- can we handle that case more gracefully? Potentially with a `configurationError` and let the user know they have to configure that prop manually due to lack of permissions in SG's settings?

**Comment by michelle0927** [comment]
@dannyroosevelt Updated to display a `configurationError`.
Screen Shot 2022-05-20 at 3 04 32 PM

**Comment by dannyroosevelt** [comment]
> @dannyroosevelt Updated to display a `configurationError`. Screen Shot 2022-05-20 at 3 04 32 PM

Excellent -- thank you!

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[SendGrid_1248_1950.pdf](https://github.com/PipedreamHQ/pipedream/files/8762765/SendGrid_1248_1950.pdf)

**Pull request merged**